File: src/main/java/com/turleylabs/algo/trader/kata/framework/BaseAlgorithm.java
package com.turleylabs.algo.trader.kata.framework;

import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public abstract class BaseAlgorithm {

    ArrayList<SimpleMovingAverage> movingAverages = new ArrayList<>();
    private LocalDate currentDate;
    private int cash;
    private LocalDate startDate;
    private LocalDate endDate;
    protected Map<String, Holding> portfolio = new HashMap<>();
    protected ArrayList<Trade> trades = new ArrayList<>();
    private Slice currentSlice;

    public BaseAlgorithm() {
        initialize();
    }

    protected abstract void initialize();


    public void processData(Slice data){
        if(data.get("TQQQ") != null) {
            currentDate = data.getDate();
            currentSlice = data;

            for(SimpleMovingAverage movingAverage : movingAverages) {
                movingAverage.addData(data.get(movingAverage.getSymbol()).getPrice());
            }

            this.onData(data);
        }
    }

    protected abstract void onData(Slice data);

    protected void setStartDate(int i, int i1, int i2) {
        this.startDate = LocalDate.of(i, i1, i2);
    }

    protected void setEndDate(int i, int i1, int i2) {
        this.endDate = LocalDate.of(i, i1, i2);
    }

    protected void setCash(int cash) {
        this.cash = cash;
    }

    public int getCash() {
        return cash;
    }

    protected LocalDate getDate() {
        return currentDate;
    }

    protected SimpleMovingAverage SMA(String symbol, int i) {
        SimpleMovingAverage movingAverage = new SimpleMovingAverage(symbol, i);
        this.movingAverages.add(movingAverage);
        return movingAverage;
    }

    protected void log(String log) {
        System.out.println(String.format("%tF - %s", currentDate, log));
    }

    protected void setHoldings(String symbol, double amount) {
        double averagePrice = currentSlice.get(symbol).getPrice();
        int shares = (int)((amount * this.getCash()) / averagePrice);
        Trade trade = new Trade(symbol, currentDate, shares, averagePrice, 1);
        trades.add(trade);
        portfolio.put(symbol, new Holding(averagePrice, shares));
        this.cash -= (averagePrice * shares);
    }

    protected void liquidate(String symbol) {
        Holding holding = portfolio.get(symbol);
        int shares = holding.getQuantity();
        double currentPrice = currentSlice.get(symbol).getPrice();
        Trade trade = new Trade(symbol, currentDate, shares, currentPrice, -1);
        trades.add(trade);
        portfolio.remove(symbol);
        this.cash += (currentPrice * shares);
    }

    public void run() {
        datesUntil(startDate, endDate).forEach(date -> processData(new Slice(date)));
    }

    public static List<LocalDate> datesUntil(LocalDate start, LocalDate end)
    {
        long diffInDays = ChronoUnit.DAYS.between(start, end);
        return Stream.iterate(start, date -> date.plusDays(1))
                .limit(diffInDays)
                .collect(Collectors.toList());
    }

}

File: typescript/src/framework/base_algorithm.ts
import { SimpleMovingAverage } from './simple_moving_average';
import { Holding } from './holding';
import { Trade } from './trade';
import { Slice } from './slice';

export abstract class BaseAlgorithm {
  private movingAverages: SimpleMovingAverage[] = [];
  private currentDate: Date;
  private cash: number;
  private startDate: Date;
  private endDate: Date;
  protected portfolio: { [key: string]: Holding } = {};
  protected trades: Trade[] = [];
  private currentSlice: Slice;

  protected abstract initialize();

  private processData(data: Slice) {
    this.currentDate = data.date;
    this.currentSlice = data;

    for (const movingAverage of this.movingAverages) {
      if (data.containsKey(movingAverage.symbol)) {
        const bar = data.get(movingAverage.symbol);
        if (bar != null) {
          movingAverage.addData(bar.price);
        }
      }
    }

    this.onData(data);
  }

  protected abstract onData(data: Slice);

  protected setStartDate(i: number, i1: number, i2: number) {
    this.startDate = new Date(Date.UTC(i, i1 - 1, i2));
  }

  protected setEndDate(i: number, i1: number, i2: number) {
    this.endDate = new Date(Date.UTC(i, i1 -1, i2));
  }

  protected setCash(cash: number) {
    this.cash = cash;
  }

  getCash(): number {
    return this.cash;
  }

  protected getDate(): Date {
    return this.currentDate;
  }

  protected SMA(symbol: string, i: number): SimpleMovingAverage {
    const movingAverage = new SimpleMovingAverage(symbol, i);
    this.movingAverages.push(movingAverage);
    return movingAverage;
  }

  protected log(log: string) {
    console.log(`${this.currentDate.toISOString().split('T')[0]} - ${log}`);
  }

  protected setHoldings(symbol: string, amount: number) {
    const averagePrice = this.currentSlice.get(symbol).price;
    const shares = Math.floor(amount * this.getCash() / averagePrice);
    const trade = new Trade(symbol, this.currentDate, shares, averagePrice, 1);
    this.trades.push(trade);
    this.portfolio[symbol] = new Holding(averagePrice, shares);
    this.cash -= (averagePrice * shares);
  }

  protected liquidate(symbol: string) {
    const holding = this.portfolio[symbol];
    const shares = holding.quantity;
    const currentPrice = this.currentSlice.get(symbol).price;
    const trade = new Trade(symbol, this.currentDate, shares, currentPrice, -1);
    this.trades.push(trade);
    delete this.portfolio[symbol];
    this.cash += currentPrice * shares;
  }

  run() {
    this.initialize();
    BaseAlgorithm.datesUntil(this.startDate, this.endDate).forEach(date => {
      this.processData(new Slice(date, 'VIX'));
      this.processData(new Slice(date, 'TQQQ'));
    });
  }
  
  static datesUntil(start: Date, end: Date): Date[] {
    const dates = [];
    let currentDate = new Date(start.getTime());
    while (currentDate.toISOString().split('T')[0] != end.toISOString().split('T')[0]) {
      dates.push(new Date(currentDate.getTime()));
      currentDate.setUTCDate(currentDate.getUTCDate() + 1);
    }
    return dates;
  }
}